### ISSUE TITLE
关于解压缩之后文件的效果失真

### ISSUE THREAD

**Issue by Alsvidermoon** [YihangChen-ee/FCGS#9]
![Image](https://github.com/user-attachments/assets/04623e3d-f973-4b23-b092-4c442c40ee6a)
![Image](https://github.com/user-attachments/assets/b430aa30-ed14-432c-ae81-e9aa7d68957c)

当我按照readme的步骤，布置完整个项目后，我用手上的ply文件做了一个测试，是一只鞋的ply，大概17万个高斯点，按照默认设置进行压缩，文件从35M压缩到了4M然后按照默认设置解码，最后解码的ply文件与原始的ply文件大小差不多，但是颜色失真比较严重，如果我想要一个较为理想的解码ply，我需要修改哪些参数，或者要进行哪些额外的操作？ @YihangChen-ee

**Comment by YihangChen-ee** [comment]
hi, 如果要保证最高的保真度，建议使用1e-4的模型。另外，建议检查一下运行环境，看看是否一致，这也可能导致潜在的误差。

**Comment by Alsvidermoon** [comment]
> hi, 如果要保证最高的保真度，建议使用1e-4的模型。另外，建议检查一下运行环境，看看是否一致，这也可能导致潜在的误差。

您好，我刚刚用另外一个ply做验证，最后解码的效果很好。原来解码不好的ply由[PGSR](https://github.com/zju3dv/PGSR)生成的，解码效果好的ply由[Pixel-GS](https://github.com/zhengzhang01/Pixel-GS)生成，查看器为supersplat，我简单的看了两个ply的内容，它们的头文件都是一样的且内容都是鞋子，只是前者为35M，17万个点，后者为78M，28万个点。具体的字段数值没有仔细看，请问您的项目对原始ply文件有什么要求吗？

**Comment by YihangChen-ee** [comment]
FCGS是使用从原始3DGS方法中optimize得到的3DGS场景进行训练的，所以当应用在其他来源的3DGS场景，可能会由于domain gap导致效果变差。你可以尝试手动将mask全部设置为0来尽可能规避这个问题。具体可以参考[Appendix](https://openreview.net/attachment?id=DCandSZ2F1&name=supplementary_material)的Section B: ANALYSIS OF LIMITATIONS。

**Comment by Alsvidermoon** [comment]
> FCGS是使用从原始3DGS方法中optimize得到的3DGS场景进行训练的，所以当应用在其他来源的3DGS场景，可能会由于domain gap导致效果变差。你可以尝试手动将mask全部设置为0来尽可能规避这个问题。具体可以参考[Appendix](https://openreview.net/attachment?id=DCandSZ2F1&name=supplementary_material)的Section B: ANALYSIS OF LIMITATIONS。

请问在哪个地方修改

**Comment by Alsvidermoon** [comment]
> FCGS是使用从原始3DGS方法中optimize得到的3DGS场景进行训练的，所以当应用在其他来源的3DGS场景，可能会由于domain gap导致效果变差。你可以尝试手动将mask全部设置为0来尽可能规避这个问题。具体可以参考[Appendix](https://openreview.net/attachment?id=DCandSZ2F1&name=supplementary_material)的Section B: ANALYSIS OF LIMITATIONS。

您好，按照您的建议将mask置零，这个问题已经解决了，目前测试了三种不同GS算法输出的ply文件，都能做到快速的编码解码。在修改过程中我有个疑问，
1.compress与compress_only函数是适用什么情况的呢？
2.能在浏览器端进行解码操作吗？

目前我的修改位置如下：
      compress函数中
        #1.mask置零
        mask_new = torch.zeros_like(mask)
        mask = mask_new + (mask - mask.detach())

        print(type(mask))
        bits_mask = encoder(
            x=mask,
            file_name=os.path.join(root_path, 'mask.b')

        )

        print('Start compressing fea...')
        #2.注释掉
        # bits_fea_hyp = encoder_factorized_chunk(
        #     x=g_fea_enc_q_hyp_q,  # [N_fea, 64]
        #     lower_func=self.EF_fea._logits_cumulative,
        #     Q=self.Q,
        #     file_name=os.path.join(root_path, 'g_fea_enc_q_hyp_q.b')
        # )

        # bits_fea_main = 0
        # for l_sp in range(4):
        #     k_st = kn[l_sp]
        #     k_ed = kn[l_sp+1]
        #     mean_sp = mean_sp_list[l_sp]
        #     scale_sp = scale_sp_list[l_sp]
        #     prob_sp = prob_sp_list[l_sp]
        #     for l_ch in range(4):
        #         c_st = l_ch*64
        #         c_ed = l_ch*64+64
        #         # from: mean_sp: [N_fea, 256], scale_sp: [N_fea, 256], prob_sp: [N_fea, 256]
        #         # to: mean_sp_l: [k_len, 64], scale_sp_l: [k_len, 64], prob_sp_l: [k_len, 64]
        #         mean_sp_l, scale_sp_l, prob_sp_l = mean_sp[:, c_st:c_ed], scale_sp[:, c_st:c_ed], prob_sp[:, c_st:c_ed]
        #         mean_hp_l, scale_hp_l, prob_hp_l = mean_hp[k_st:k_ed, c_st:c_ed], scale_hp[k_st:k_ed, c_st:c_ed], prob_hp[k_st:k_ed, c_st:c_ed]
        #         mean_ch_l, scale_ch_l, prob_ch_l = mean_ch[k_st:k_ed, c_st:c_ed], scale_ch[k_st:k_ed, c_st:c_ed], prob_ch[k_st:k_ed, c_st:c_ed]
        #         probs = torch.stack([prob_sp_l, prob_hp_l, prob_ch_l], dim=-1)  # [k_len, 256, 3]
        #         probs = torch.softmax(probs, dim=-1)  # [k_len, 256, 3]
        #         prob_sp_l, prob_hp_l, prob_ch_l = probs[..., 0], probs[..., 1], probs[..., 2]  # [k_len, 256] for each
        #         bits_fea_main_tmp = encoder_gaussian_mixed_chunk(
        #             x=g_fea_enc_q[k_st:k_ed, c_st:c_ed].contiguous().view(-1),  # from: [N_fea, 256]
        #             mean_list=[mean_sp_l.contiguous().view(-1), mean_hp_l.contiguous().view(-1), mean_ch_l.contiguous().view(-1)],
        #             scale_list=[scale_sp_l.contiguous().view(-1), scale_hp_l.contiguous().view(-1), scale_ch_l.contiguous().view(-1)],
        #             prob_list=[prob_sp_l.contiguous().view(-1), prob_hp_l.contiguous().view(-1), prob_ch_l.contiguous().view(-1)],
        #             Q=self.Q,
        #             file_name=os.path.join(root_path, f'g_fea_enc_q_sp{l_sp}_ch{l_ch}.b'),
        #             chunk_size=c_size_fea,
        #         )
        #         bits_fea_main += bits_fea_main_tmp
        # bits_fea = bits_fea_main + bits_fea_hyp
        #3.添加
        bits_fea_main = 0
        bits_fea_hyp = 0
        bits_fea = 0
        print('Start compressing feq...')

        在decompress函数中：
         print('Start decompressing fea...')
        # 1.注释
        # g_fea_enc_q_hyp_q = decoder_factorized_chunk(
        #     lower_func=self.EF_fea._logits_cumulative,
        #     Q=self.Q,
        #     N_len=mask_fea.sum().item(),  # N_fea
        #     dim=64,
        #     file_name=os.path.join(root_path, 'g_fea_enc_q_hyp_q.b')
        # )  # [N_fea, 64]
        # mean_hp, scale_hp, prob_hp = torch.split(self.Decoder_fea_hyp(g_fea_enc_q_hyp_q), split_size_or_sections=[256, 256, 256], dim=-1)  # [N_fea, 256] for each

        # g_fea_enc_q = torch.zeros(size=[mask_fea.sum(), 256], dtype=torch.float32, device='cuda')  # [N_fea, 256]
        # for l_sp in range(4):
        #     k_st = kn[l_sp]
        #     k_ed = kn[l_sp+1]
        #     s_st = sn[l_sp]
        #     s_ed = sn[l_sp+1]
        #     if l_sp == 0:
        #         ctx_sn = torch.zeros(size=[k1-k0, self.cafea_indim-self.freq_enc.output_dim], dtype=torch.float32, device='cuda')  # [k_st-k0, 48]=[k1-k0, 48]
        #     else:
        #         fea_grid_feature_curr = self.latdim_2_griddim_fea(g_fea_enc_q[k0:k_st])  # [k_st-k0, 48]
        #         # norm_xyz_clamp: [N_g, 3], mask_fea: [N_g], fea_grid_feature_curr: [k_st-k0, 48], norm_xyz_clamp[s_st:s_ed][mask_fea[s_st:s_ed]]: [k_ed-k_st, 3]
        #         ctx_sn = self.feq_spatial_ctx.forward(norm_xyz_clamp[s0:s_st][mask_fea[s0:s_st]], norm_xyz_clamp[s_st:s_ed][mask_fea[s_st:s_ed]], fea_grid_feature_curr, determ=True)  # [k_ed-k_st, dim]
        #     ctx_sn = torch.cat([ctx_sn, freq_enc_xyz[s_st:s_ed][mask_fea[s_st:s_ed]]], dim=-1)  # [k_ed-k_st, dim]
        #     mean_sp, scale_sp, prob_sp = torch.split(self.context_analyzer_fea(ctx_sn), split_size_or_sections=[256, 256, 256], dim=-1)
        #     for l_ch in range(4):
        #         c_st = l_ch*64
        #         c_ed = l_ch*64+64
        #         mean_sp_l, scale_sp_l, prob_sp_l = mean_sp[:, c_st:c_ed], scale_sp[:, c_st:c_ed], prob_sp[:, c_st:c_ed]  # [k_ed-k_st, 48] for each
        #         mean_hp_l, scale_hp_l, prob_hp_l = mean_hp[k_st:k_ed, c_st:c_ed], scale_hp[k_st:k_ed, c_st:c_ed], prob_hp[k_st:k_ed, c_st:c_ed]  # [k_ed-k_st, 48] for each
        #         mean_ch_l, scale_ch_l, prob_ch_l = self.fea_channel_ctx.forward(g_fea_enc_q[k_st:k_ed], to_dec=l_ch)  # [k_ed-k_st, 48] for each
        #         probs = torch.stack([prob_sp_l, prob_hp_l, prob_ch_l], dim=-1)  # [N_fea, 48, 3]
        #         probs = torch.softmax(probs, dim=-1)  # [N_fea, 48, 3]
        #         prob_sp_l, prob_hp_l, prob_ch_l = probs[..., 0], probs[..., 1], probs[..., 2]  # [N_fea, 48] for each
        #         g_fea_enc_q_tmp = decoder_gaussian_mixed_chunk(
        #             mean_list=[mean_sp_l.contiguous().view(-1), mean_hp_l.contiguous().view(-1), mean_ch_l.contiguous().view(-1)],
        #             scale_list=[scale_sp_l.contiguous().view(-1), scale_hp_l.contiguous().view(-1), scale_ch_l.contiguous().view(-1)],
        #             prob_list=[prob_sp_l.contiguous().view(-1), prob_hp_l.contiguous().view(-1), prob_ch_l.contiguous().view(-1)],
        #             Q=self.Q,
        #             file_name=os.path.join(root_path, f'g_fea_enc_q_sp{l_sp}_ch{l_ch}.b'),
        #             chunk_size=c_size_fea,
        #         ).contiguous().view(k_ed-k_st, c_ed-c_st)  # [k_ed-k_st, 64]
        #         g_fea_enc_q[k_st:k_ed, c_st:c_ed] = g_fea_enc_q_tmp

        # g_fea_out = self.Decoder_fea(g_fea_enc_q)  # [N_fea, 256]
        # fe_dec = torch.cat([self.head_f_dc(g_fea_out), self.head_f_rst(g_fea_out)], dim=-1)  # [N_fea, 48]

       # 2.注释
        fe_q = torch.zeros(size=[mask_feq.sum(), 48], dtype=torch.float32, device='cuda')  # [N_feq, 48]
        fe_final = torch.zeros(size=[N_g, 48], dtype=torch.float32, device='cuda')  # [N_g, 48]
        #fe_final[mask_fea] = fe_dec
        for l_sp in range(4):
            t_st = tn[l_sp]
            t_ed = tn[l_sp+1]
            s_st = sn[l_sp]
            s_ed = sn[l_sp+1]

**Comment by YihangChen-ee** [comment]
你好，

1. 使用compress函数就行了。compress_only的话有一些比较迷的问题（可能是pytorch的精度导致的），导致解码会有点不匹配。两者的原理没有任何差别。后续版本会把compress_only这个函数删了。
2. 浏览器的适配是工程的问题，而我们目前仅限于算法的研究。你可以自己尝试适配一下，理论上是可以支持的。

BTW：你修改的代码没有问题，不过建议可以直接定义新的mask： mask = torch.zeros_like(mask)，不需要通过mask_new作为中介。

祝好

**Comment by Lucas-XJE** [comment]
> ![Image](https://github.com/user-attachments/assets/04623e3d-f973-4b23-b092-4c442c40ee6a) ![Image](https://github.com/user-attachments/assets/b430aa30-ed14-432c-ae81-e9aa7d68957c)
> 
> 当我按照readme的步骤，搭建完整个之后，我用手上的ply文件做了一个测试，是一只鞋的ply，大概17万个高点，按照默认设置进行压缩，文件从35M压缩到了4M然后按照默认设置解码，最后解码的ply文件与原始的ply文件大小差不多，但是颜色失真比较严重，如果我想要一个不太理想的解码，我需要修改哪些参数，或者要进行哪些额外的操作？[@YihangChen-ee](https://github.com/YihangChen-ee)

@Alsvidermoon 您好，请问这只鞋子的数据集和重建的程序是什么

**Comment by Lucas-XJE** [comment]
> > > ![Image](https://github.com/user-attachments/assets/04623e3d-f973-4b23-b092-4c442c40ee6a) ![图像](https://private-user-images.githubusercontent.com/115556514/417101861-b430aa30-ed14-432c-ae81-e9aa7d68957c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA4ODQ3MDUsIm5iZiI6MTc2MDg4NDQwNSwicGF0aCI6Ii8xMTU1NTY1MTQvNDE3MTAxODYxLWI0MzBhYTMwLWVkMTQtNDMyYy1hZTgxLWU5YWE3ZDY4OTU3Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAxOVQxNDMzMjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNTA3Mzg1NzZhOWU2N2E4N2E4ZDRkYzYxNjkxODYwMjY3MmU5MGU3NWE3NWJiODUzZDhkNzA4N2M3ZTYxZDAwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vvjf51VfANPcM2cux8QeFVKk0NGDIEdET1-YZ4eGXmE)
> > > 当我readme的步骤，搭建完整时，我用手上的ply文件做了一个要测试，是一只鞋的ply，大约17万高点，按照默认设置进行压缩，文件从35M压缩到了4M然后默认按照设置解码，最后解码的ply文件与原始的ply文件大小差不多，但是颜色不太理想之后的解码，如果我想要一个不太理想的解码，我需要修改哪些参数，或者进行哪些额外的操作？[@YihangChen-ee](https://github.com/YihangChen-ee)
> > 
> > 
> > [@Alsvidermoon](https://github.com/Alsvidermoon)您好，请问这只鞋子的数据集和重建的程序是什么
> 
> 鞋子是我自己拍的，用pgsr重建的

谢谢你
